{"cells":[{"cell_type":"markdown","metadata":{},"source":["Group Name: AG xx.\n","\n","Student Name (Student ID):\n","\n","1. xxxx xxxxx (xxxxxxx)\n","\n","2. xxxx xxxxx (xxxxxxx)\n","\n","3. xxxx xxxxx (xxxxxxx)"]},{"cell_type":"markdown","metadata":{"id":"yZOz6Y_cLGOT"},"source":["# Question 2\n","\n","## Introduction to question 2\n","\n","In the second question of this assignment, we will explore the use of local search in genome assembly.\n","\n","We will use local search to assemble (construct) a large part of the nucleotide sequence of the monkeypox virus, which has been downloaded from the National Center for Biotechnology Information in the United States. Please note that no additional or specialized knowledge of biology or bioinformatics is required for this assignment. (Actually, the technical specifics of bioinformatics have been adapted and simplified for the purposes of this computer science assignment, so if you are a biologist, please do not apply preexisting knowledge to solve the problem. Furthermore, you should not attempt to search up the genome on genomic databases to \"guess\" the actual sequence, since we are more interested in your coding methodology rather than your attempts at reproducing a known sequence.)\n","\n","This is an introductory computer science assignment and not a bioinformatics assignment; we are simply using bioinformatics as a use case to illustrate the applicability of local search to the natural sciences. Therefore, no knowledge of bioinformatics is assumed or required. In the paragraphs that follow, I will give a short crash course which will cover all the domain knowledge you will need to know in order to tackle this problem.  \n","\n","For technical reasons, when we analyze the nucleotide sequence (genome) of a virus, we usually cannot “read” it in one fell swoop. We have to read the genome in parts, because the genome is usually too long for the machine to read in a single sitting. To simplify things, a “read” is a single view of part of the genome; think of it as a SUBSTRING, a partial view of the whole genome. After we have generated multiple reads of a genome, we then have to “stitch”, or combine, the different reads of the genome together. This process of stitching up reads of a genome into the final sequence is known as genome assembly. However, the different reads of the genome cannot just be concatenated like usual string concatenation. It’s not a situation where you have one read, “Hello”, and another read, “World”, and all you need to do is concatenate both strings together to make “Hello World”. Among other reasons, there are two major reasons why you can’t do so:\n","\n","1. You do not know which read came first. The reads are not ordered. How do you know “Hello” came after “World”? The answer is that you don’t. Imagine how complicated this situation might be if you had more than two reads. (This is indeed our situation, where we have $n$ reads, and $n>>2$.)\n","\n","2. One read may contain a substring contained in another read. Specifically, without loss of generality, part of the ending $x$ characters of a read (i.e., suffix) might also be found in the starting $x$ positions (i.e., prefix) of another read.\n","\n","- A computer scientist usually creates opportunities from problems. While this may be a “problem” in that you just can’t concatenate two strings blindly, the fact that strings contain shared “substrings” is actually a very helpful clue that you can use to “join” strings together. \n","\n","- Note that the choice of the value of $x$ could be a hyperparameter decided by the computer scientist.\n","\n","## Your tasks\n","\n","In this part of the assignment, you will work with (simulated) reads that I have generated from the nucleotide sequence of the monkeypox virus. In reality, bioinformatics is far more complicated, but here we will work with a simplified situation. Your task is to examine the reads that I have provided for you, and from there “infer” the nucleotide sequence that might have produced those reads. \n","\n","The reads are provided in the csv file `data.csv` which simply provides a list of unique strings. Note that you should NOT assume any particular ordering of the strings in this dataframe. In fact, the strings have already been shuffled randomly. \n","\n","NOTE: You are not allowed to use `pandas` or any other libraries apart from the Python STL to load the csv file.\n","\n","### Task A (3 marks): \n","\n","Create a directed graph. The nodes in the graph are the strings in the list of reads. An edge should be drawn FROM read A TO read B if and only if a suffix (of length $x$) of read A is also a prefix (obviously, also of length $x$) of read B. For the purposes of the assignment, limit the value of $x$ to between 5 and 30, both inclusive. That is, to be clear, $5\\leq x\\leq 30$. The weight of an edge between read A and read B should be the NEGATED value of $x$, i.e. $-x$. \n","\n","In your Jupyter notebook, please report the number of edges in your graph. Provide a barplot or histogram which shows the number of edges with different weights or weight categories. In this task, you are free to use plotting libraries such as `matplotlib` or `seaborn` to plot this graph.\n","\n","As an example, if read A is \"TACTAGT\" and read B is \"TAGTCCCCT\", then an edge is drawn FROM read A TO read B (i.e., $A \\rightarrow B$) with weight of $-4$. This is because the 4-suffix \"TAGT\" is also the 4-prefix of read B; in other words, the last 4 characters of read A (a substring of length 4) overlap with the first 4 characters of read B (a substring of length 4).\n","\n","### Task B (7 marks): \n","\n","From Task A, you now have a graph which shows connections between reads based on how they overlap, in theory you could draw a path through the graph and thereby derive the full sequence (genome).\n","\n","Task B asks you to use local search method(s) to determine a path through this directed graph of strings. \n","\n","- You are expected to use simulated annealing and tune the relevant configuration settings and hyperparameters. The minimum requirement is to implement simulated annealing.\n","\n","- Explain tha rationale behind the choice of scheduling strategy and parameters.\n","\n","- However, you may also explore other search methods in addition to simulated annealing. Marks will be awarded for effort.\n","\n","Note the following constraints:\n","\n","1. The path has to go through each and every vertex exactly once. For computer scientists, this constraint is reminiscent of the \"Traveling Salesman's Problem\", except that unlike TSP, we should not need to go back to the starting vertex again. \n","\n","2. For the purposes of neighbor generation / action selection at each node, bear in mind that a path through the graph which minimizes the total number of nucleotides in the assembled sequence is the preferred path. To state that another way, the assembled sequence should be derived from a path that goes through EACH and EVERY vertex exactly once, however we want this assembled sequence to be AS SHORT AS POSSIBLE.\n","\n","3. You are not given the starting (source/origin) or ending (destination) vertex.\n","\n","4. For avoidance of ambiguity, no cycles are allowed. You must not visit a vertex more than once.\n","\n","5. You are not allowed to use any libraries apart from the Python Standard Library.\n","No import statements which import libraries outside of the Python STL should be found within your answer for Task B.\n","\n","Please remember to report the assembled sequence that you obtain. Although it would be great if you can come up with a good sequence, please feel reassured that we are more interested in your APPROACH to the problem, and so you can potentially get a reasonable score on this task even if your solution is \"wrong\". It is the process, rather than the result, which matters more."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"oBnbZpd2bt1N"},"outputs":[],"source":["# Problem Class\n","class Problem:\n","    \"\"\"The abstract class for a formal problem. A new domain subclasses this,\n","    overriding `actions` and `results`, and perhaps other methods.\n","    The default heuristic is 0 and the default action cost is 1 for all states.\n","    When you create an instance of a subclass, specify `initial`, and `goal` states \n","    (or give an `is_goal` method) and perhaps other keyword args for the subclass.\"\"\"\n","\n","    def __init__(self, initial=None, goal=None, **kwds): \n","        self.__dict__.update(initial=initial, goal=goal, **kwds) \n","        \n","    def actions(self, state):        raise NotImplementedError\n","    def result(self, state, action): raise NotImplementedError\n","    def is_goal(self, state):        return state == self.goal\n","    def action_cost(self, s, a, s1): return 1\n","    def h(self, node):               return 0\n","    \n","    def __str__(self):\n","        return '{}({!r}, {!r})'.format(\n","            type(self).__name__, self.initial, self.goal)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Use the following Node class to generate search tree\n","import math\n","class Node:\n","    \"A Node in a search tree.\"\n","    def __init__(self, state, parent=None, action=None, path_cost=0):\n","        self.__dict__.update(state=state, parent=parent, action=action, path_cost=path_cost)\n","\n","    def __repr__(self): return '<{}>'.format(self.state)\n","    def __len__(self): return 0 if self.parent is None else (1 + len(self.parent))\n","    def __lt__(self, other): return self.path_cost < other.path_cost "]},{"cell_type":"code","execution_count":null,"metadata":{"id":"B2-y3Nt9ccxj"},"outputs":[],"source":["# Code to generate neighbours, value of states, etc.\n","class TSP(Problem):\n","    #Implement TSP class here\n","    pass"]}],"metadata":{"colab":{"collapsed_sections":[],"provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}
